cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/MinTopicLeadersPerBrokerGoal.java

```diff
@@ -158,9 +160,16 @@ private boolean doesLeaderRemoveViolateOptimizedGoal(Replica replicaToBeRemoved)
   @Override
   protected void initGoalState(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
       throws OptimizationFailureException {
-    _mustHaveTopicLeadersPerBroker = Collections.unmodifiableSet(
+    Set<String> mustHaveTopicLeadersPerBroker = Collections.unmodifiableSet(
         Utils.getTopicNamesMatchedWithPattern(_balancingConstraint.topicsWithMinLeadersPerBrokerPattern(), clusterModel::topics));
-    if (_mustHaveTopicLeadersPerBroker.isEmpty()) {
+    Map<String, Integer> numLeadersByTopicNames = clusterModel.numLeadersPerTopic(mustHaveTopicLeadersPerBroker);
+    int eligibleBrokersForLeadership = eligibleBrokersForLeadership(clusterModel, optimizationOptions).size();
+    _mustHaveTopicMinLeadersPerBroker = new HashMap<>();
+    for (String topicName : mustHaveTopicLeadersPerBroker) {
+      int topicNumLeaders = numLeadersByTopicNames.get(topicName);
+      _mustHaveTopicMinLeadersPerBroker.put(topicName, eligibleBrokersForLeadership == 0 ? 0 : topicNumLeaders / eligibleBrokersForLeadership);
+    }
```

Comment:
If `mustHaveTopicLeadersPerBroker` is empty, this method will still do a significant work before returning.
Similarly, calls to `clusterModel.numLeadersPerTopic(mustHaveTopicLeadersPerBroker)` and `eligibleBrokersForLeadership` are only relevant if the `min.topic.leaders.per.broker` is non-positive.

Should we avoid introducing extra work when possible?

Reply:
Good call,  updated to compute the map conditionally